src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Windows.cs

```diff
@@ -0,0 +1,272 @@
+            {
+                FillKeyDerivation(password, salt, iterations, hashAlgorithmName.Name, destination);
+            }
+            else
+            {
+                FillDeriveKeyPBKDF2(password, salt, iterations, hashAlgorithmName.Name, destination);
+            }
+        }
+
+        private static unsafe void FillKeyDerivation(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterations,
+            string hashAlgorithmName,
+            Span<byte> destination)
+        {
+            SafeBCryptKeyHandle keyHandle;
+            int hashBlockSizeBytes = GetHashBlockSize(hashAlgorithmName);
+
+            // stackalloc 0 to let compiler know this cannot escape.
+            Span<byte> clearSpan = stackalloc byte[0];
+            ReadOnlySpan<byte> symmetricKeyMaterial = stackalloc byte[0];
+            int symmetricKeyMaterialLength;
+
+            if (password.IsEmpty)
+            {
+                // CNG won't accept a null pointer for the password.
+                symmetricKeyMaterial = stackalloc byte[1];
+                symmetricKeyMaterialLength = 0;
+                clearSpan = default;
+            }
+            else if (password.Length <= hashBlockSizeBytes)
+            {
+                // Password is small enough to use as-is.
+                symmetricKeyMaterial = password;
+                symmetricKeyMaterialLength = password.Length;
+                clearSpan = default;
+            }
+            else
+            {
+                // RFC 2104: "The key for HMAC can be of any length (keys longer than B bytes are
+                //     first hashed using H).
+                //     We denote by B the byte-length of such
+                //     blocks (B=64 for all the above mentioned examples of hash functions)
+                //
+                // Windows' PBKDF2 will do this up to a point. To ensure we accept arbitrary inputs for
+                // PBKDF2, we do the hashing ourselves.
+                Span<byte> hashBuffer = stackalloc byte[512 / 8]; // 64 bytes is SHA512, the largest digest handled.
+                int hashBufferSize;
+
+                switch (hashAlgorithmName)
+                {
+                    case HashAlgorithmNames.SHA1:
+                        hashBufferSize = SHA1.HashData(password, hashBuffer);
+                        break;
+                    case HashAlgorithmNames.SHA256:
+                        hashBufferSize = SHA256.HashData(password, hashBuffer);
+                        break;
+                    case HashAlgorithmNames.SHA384:
+                        hashBufferSize = SHA384.HashData(password, hashBuffer);
+                        break;
+                    case HashAlgorithmNames.SHA512:
+                        hashBufferSize = SHA512.HashData(password, hashBuffer);
+                        break;
+                    default:
+                        throw new CryptographicException();
```

Comment:
Hash algorithm selection was already filtered, right?  Let's throw in a Debug.Fail that reminds readers of that, so it's not just a "huh, weird, a default-message throw".